examples/bloom-embeddings.ipynb

```diff
@@ -0,0 +1,729 @@
+    "    key1 = mmh3.hash(word, 0) % 15\n",
+    "    key2 = mmh3.hash(word, 1) % 15\n",
+    "    vector = hashed_embed[key1] + hashed_embed[key2]\n",
+    "    print(word, '%.3f %.3f' % tuple(vector))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We now have a function that maps our 20 words to 20 unique vectors – but we're\n",
+    "storing weights for only 15 vectors in memory. Now the question is: will we be\n",
+    "able to find values for these weights that let us actually map words to useful\n",
+    "vectors?\n",
+    "\n",
+    "Let's do a quick experiment to see how this works. We'll assign \"true\" values\n",
+    "for our little vocabulary, and see how well we can approximate them with our\n",
+    "compressed table. To get the \"true\" values, we _could_ put the \"science\" in data\n",
+    "science, and drag the words around into reasonable-looking clusters. But for our\n",
+    "purposes, the actual \"true\" values don't matter. We'll therefore just do a\n",
+    "simulation: we'll assign random vectors as the \"true\" state, and see if we can\n",
+    "learn values for the hash embeddings that match them.\n",
+    "\n",
+    "The learning procedure will be a simple stochastic gradient descent:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {
+    "colab": {
+     "background_save": true
+    },
+    "executionInfo": {
+     "elapsed": 3,
+     "status": "aborted",
+     "timestamp": 1648199186370,
+     "user": {
+      "displayName": "Vincent D. Warmerdam",
+      "photoUrl": "https://lh3.googleusercontent.com/a-/AOh14Gh4KYzhhhK0YDTnAQsUIaQPw-0dKIP-kLBID7nFdQ=s64",
+      "userId": "05641618555626735638"
+     },
+     "user_tz": -60
+    },
+    "id": "ET4n9AA5y0fX"
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "499 43.47128495286\n"
+     ]
+    }
+   ],
+   "source": [
+    "import numpy\n",
+    "import numpy.random as random\n",
+    "import mmh3\n",
+    "\n",
+    "random.seed(0)\n",
+    "nb_epoch = 500\n",
+    "learn_rate = 0.001\n",
+    "nr_hash_vector = 1000\n",
+    "\n",
+    "words = [str(i) for i in range(2000)]\n",
+    "true_vectors = numpy.random.uniform(-0.1, 0.1, (len(words), 10))\n",
+    "hash_vectors = numpy.random.uniform(-0.1, 0.1, (nr_hash_vector, 10))\n",
+    "examples = list(zip(words, true_vectors))\n",
+    "\n",
+    "for epoch in range(nb_epoch):\n",
+    "    random.shuffle(examples)\n",
+    "    loss=0.\n",
+    "    for word, truth in examples:\n",
+    "        key1 = mmh3.hash(word, 0) % nr_hash_vector\n",
+    "        key2 = mmh3.hash(word, 1) % nr_hash_vector\n",
+    "        hash_vector = hash_vectors[key1] + hash_vectors[key2]\n",
+    "        diff = hash_vector - truth\n",
+    "        hash_vectors[key1] -= learn_rate * diff\n",
+    "        hash_vectors[key2] -= learn_rate * diff\n",
+    "        loss += (diff**2).sum()\n",
```

Comment:
I'm always very confused with derivatives, but here the loss is the `(x - y)^2` and the derivative is `x - y`. It doesn't really matter I guess but the if the derivative is `x - y` then the loss is `1/2 * (x - y)^2` right?

Reply:
I don't think this really matters, and as far as I can tell it's what's currently in e.g. `CategoricalCrossentropy`.

Reply:
but this is the sum of squared error loss not cross-entropy. The gradient is the same.

Reply:
Also I believe that the loss shown in the categorical cross entropy is wrong and the gradient is correct.

Reply:
Ah, I see what you're saying. Since it doesn't make a practical difference for this demo, I think I would just leave it the way Matt wrote it for now, but we'll know what they're referring to if anyone brings it up.